File: easy/trees/python/98-validate-bst.py
from typing import Optional
import math


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.is_valid_bst_helper(root, -math.inf, math.inf)

    def is_valid_bst_helper(self, root: Optional[TreeNode], lower, upper):
        if not root:
            return True

        if lower < root.val < upper:
            return self.is_valid_bst_helper(root.left, lower, root.val) and self.is_valid_bst_helper(root.right, root.val, upper)
        else:
            return False

File: easy/trees/cpp/98-validate-bst.cpp
#include "../../../stdc++.h"
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution
{
public:
    bool isValidBST(TreeNode *root)
    {
        return bst_helper(root, LONG_MIN, LONG_MAX);
    }
    bool bst_helper(TreeNode *node, long lower, long upper)
    {
        if (node == nullptr)
        {
            return true;
        }
        if (lower < node->val && node->val < upper)
        {
            return bst_helper(node->left, lower, node->val) && bst_helper(node->right, node->val, upper);
        }
        return false;
    }
};
